Codebase rewrite

farsil opened the pull request:
I have rewritten `rofication` almost from scratch. 

### Features of the rewritten codebase

- Ditched `jsonpickle` in favor of the python standard JSON serializer. This would technically allow easier interoperability with the server and also gets rid of a dependence!
- **Improved compliance with the [Desktop Notification Specification](https://people.gnome.org/~mccann/docs/notification-spec/notification-spec-latest.html) by adding a more proper handling of expired notifications and notification ID.**
- `notify-send` notifications are always allowed to expire (useful for tests mostly).
- **Implemented the `actions` capability, it is triggered via the `see` command sent by the GUI when the `<Enter>` key is pressed. This allows applications like Firefox to interact with `rofication`**.
- `rofi` is now called prior to  `rofication-i3blocks` execution, and it is blocking. This allows the blocklet to output the updated notification count when the user stops interacting with the GUI, and also gets rid of a minor graphical glitch.
- `rofication-daemon` should be more fault-tolerant in general.
- Various efficiency improvements:
    - `rofication-i3blocks` now calls the rofication GUI in-process and not using `Popen`.
    - `rofication-i3blocks` limits calls to `xrescat` to the resources that are actually needed.
    - Command `list` now ships the serialized notification queue in one go, letting the underlying 
socket handle all the buffering.

Also, for the sake of reporting the server version to D-Bus (see [GetServerInformation](https://people.gnome.org/~mccann/docs/notification-spec/notification-spec-latest.html#command-get-server-information)) I used version number `2.0.0`, given that it is a massive revamp and it breaks backwards-compatibility due to the protocol and package changes.

### Known issues

- Sometimes, after invoking a notification action, `rofi` decides to change screen to where the application that invoked the action is.
- The `rofication-status` script (previously `rofication-i3blocks.py`) cannot be placed easily outside the `/usr/share/rofication` path since it now depends on the `rofication` package. It works if it is symlinked however.

### Tinkering information
`rofication` is unfortunately not yet ready to accept a configuration file. You may change its behavior by modifying the source code:

- Resources definitions are placed in file `rofication/resources/_static.py`. They are used in the `rofication-status` script.
- The appearance of `rofi` may be customized by editing the tuple `ROFI_COMMAND` in the `rofication/_gui.py` file.
- Adding new commands is now much easier, you have to modify the `RoficationClient` class in file `rofication/_client.py` and the `RoficationRequestHandler` class in the `rofication/_server.py` class.
- To modify the name of the serialized notification queue path, check the `rofication-daemon` script.
- To modify the name of the UNIX socket, check the `ROFICATION_UNIX_SOCK` constant in `rofication/_static.py`.


### Final remarks

- The rofication scripts have all changed names. This may or may not be desired, but in general scripts with a shebang don't need an extension (see the AWS cli, which is entirely written in python). This means that all the package references need to be updated, and that on upgrade, the previous scripts must be removed.
- The notification queue dump `not.json` should really be written in `/var/cache/rofication/` or something like that, not in the current working directory.
- We should add a `LICENSE` file in the repository, so that it is properly GPL licensed
- `rofication` was tested on an Ubuntu 19.10 machine with python 3.7. If we want to support Ubuntu 18.04 out of the box despite the incoming 20.04 LTS, the scripts will probably need to be downgraded.

farsil committed:
Implemented 'actions' capability

This required quite a rework of the notification daemon, which now
should better handle notification IDs. The daemon was also split into
several components instead of putting everything into one file.

farsil committed:
Use observer pattern to provide callbacks

farsil committed:
Code style changes

farsil committed:
Implemented proper notification expiration

farsil committed:
Notification queue externally synchronized

farsil committed:
Better names for script variables

farsil committed:
Stop using the deprecated GObject

farsil committed:
Simpler string representation

farsil committed:
Added type hinting for the daemon

More type hinting, squash me

farsil committed:
Corrected warn() handling

farsil committed:
Use serversocket instead of plain sockets

farsil committed:
Fixed regression in urgent notifications count

farsil committed:
Better type handling for IntEnum

farsil committed:
Stop exposing the underlying queue

farsil committed:
Notification is now a dataclass

farsil committed:
More fault-tolerant notification queue load/save

farsil committed:
Replaced jsonpickle with the json module

farsil committed:
Do not overlap detected last_id

farsil committed:
Readability improvements

farsil committed:
rofication-gui is now invoked in-process

farsil committed:
Organized files into rofication package

farsil committed:
Client cmds now sent with RoficationClient

farsil committed:
Move resources definitions to specific package

farsil committed:
Client/Server refactoring

farsil committed:
D-Bus service refactoring

farsil committed:
No need for a main function in pure scripts

farsil committed:
NotificationQueue now uses a dict internally

farsil committed:
More internal refactoring

farsil committed:
Scripts don't need the extension

farsil committed:
Allow notify-send notifications to expire

kgilmer commented:
Thank you for the super detailed PR notes @farsil!  this is a big refactor, and all these changes look great. :smile:   Regarding Ubuntu 18.04 support, can you provide more details about what will need to be downgraded?

farsil commented:
I have investigated the version that 18.04 uses, [which is 3.6.5](https://packages.ubuntu.com/bionic/python3), and I found out it's not too bad, the only thing I should have to refactor are the `@dataclasses` which are convenient but ultimately not necessary. I will test with a 3.6 `virtualenv` later this week and report back.

I still need to rebase my repository so that is in sync with your master and will probably add a `setup.py` file so that it's easy to install with `pip3` or similar tools. I will have to study a bit how package distribution works, it's been a while I haven't done any serious python development :)

Glad you like the changes!

farsil committed:
Downgraded required python version to 3.6

farsil committed:
Invoke RoficationGui inside try/except

farsil commented:
That was quite a complicated rebase, because I had quite a few conflicts on the commit dates. Sorry for the mess, it should be fine now.

I verified that the only thing preventing us from using the script in Ubuntu 18.04 were data classes, which I refactored. The script should be now python 3.6 compatible!

farsil committed:
Expose rofication metadata

farsil committed:
Added setup.py script

farsil committed:
Updated debian package metadata

farsil commented:
@kgilmer I updated the copyright information, and added a `setup.py` script. The script automates installation of rofication with `pip3`. Just invoke `pip3 install <path-to-rofication>` and it will place the files in the right place to be discoverable by the OS. This means two things:

- Any python application can import our `rofication` module without worrying about the filesystem path. This allows the scripts to be placed in any directory and will still function.
- The scripts are placed in python's `bin` folder, which means they can be accessed like other normal commands from anywhere. If you don't want this, I can exclude the scripts installation.

I have also modified the debian package metadata accordingly, except the changelog.

On a side note, don't you think it may be nice to have a `systemd` unit handle the daemon?

farsil commented:
I fixed a few things with the setup script and lowered the required versions of `PyGObject` and `dbus-python` so that they should match those shipped with Ubuntu 18.04.

kgilmer commented:
> On a side note, don't you think it may be nice to have a systemd unit handle the daemon?

Yes, that would be nice!  I haven't gotten much time with systemd yet...  Do you know what's involved?

kgilmer commented:
I'm attempting to host your rewrite here: https://launchpad.net/~regolith-linux/+archive/ubuntu/experimental

kgilmer commented:
A diff of changes I need to make on your branch to build the package:

```diff
$ git diff debian/control
diff --git a/debian/control b/debian/control
index 5a279ef..cc2ff08 100644
--- a/debian/control
+++ b/debian/control
@@ -2,7 +2,12 @@ Source: regolith-rofication
 Section: x11
 Priority: optional
 Maintainer: Ken Gilmer <kgilmer@gmail.com>
-Build-Depends: debhelper (>= 11)
+Build-Depends: debhelper (>= 11),
+    dh-python,
+    python-all,
+    python-setuptools,
+    python3-all,
+    python3-setuptools
 Standards-Version: 4.1.3
 Homepage: https://github.com/regolith-linux/regolith-rofication
```

farsil commented:
Strange,  `python-all` and `python-setuptools` shouldn't be required, because the script runs in Python 3. The rest make sense, I have to see if I can do without `setuptools` to minimize dependencies, I have seen a few packages do this:

```python
try:
    from setuptools import setup
except ImportError:
    from distutils.core import setup
```

farsil commented:
> 
> 
> > On a side note, don't you think it may be nice to have a systemd unit handle the daemon?
> 
> Yes, that would be nice! I haven't gotten much time with systemd yet... Do you know what's involved?

It shouldn't be too difficult, I have tinkered with unit files back when I used arch.

kgilmer commented:
Regarding the control changes, I got those packages from here: https://wiki.debian.org/Python/LibraryStyleGuide.

Regarding the dependencies, I wouldn't really worry too much about it.  They are build dependencies so have no impact on the target system, as far as I know.

kgilmer commented:
After I added the experimental PPA to a test laptop running Regolith unstable on Ubuntu 19.10 and installed this version, this is what I found:

1. The scripts fail because /usr/bin/python is not installed.  I notice on this system I have python 3.7.5, and no /usr/bin/python but I do have /usr/bin/python3
2. After manually changing the first line of the scripts to `/usr/bin/python3` , when I run rofication-status I get `ModuleNotFoundError: No module named 'rofication'`.
3. I get the same error when running `rofication-daemon`.

I'm guessing that there is something not right w/ the native python installer integration with Debian, but I don't know what it is.  I'm not too up on the latest Python packaging stuff.  If you'd like to build the debian package yourself, you can follow the setup info I have in some of the Regolith READMEs such as https://github.com/regolith-linux/regolith-desktop, and then install the `pbuilder` tool.  In essence:

```bash
$ debuild -S -sa
$ # ...generates package metadata, including changes file...
$ sudo pbuilder --build --distribution eoan ../regolith-rofication_1.2.0-1.dsc
$ # generates deb file in /var/cache/pbuilder/result or produces build error
```

kgilmer commented:
We can also try reverting to the old way of installing via `debian/install` but I guess we lose some of the Python library integration niceties.  Your call.  :smile:

farsil commented:
Hmm I have no experience with the debian build system but I can try to investigate a bit. Sure I miss arch's simple `PKGBUILD` file. I will spin an ubuntu 18.04 vm these days so I can do some tests.

Also, I think you need `debian/install` anyway, but I am not sure what to put in it.

farsil commented:
I updated the debian files, they should now work as intended! Can you please try to see if it works for you as well?

kgilmer commented:
I published eoan and bionic package versions for testing on the same PPA, and after fixing the i3 config file (it was using /usr/share...xxx.py) I was able to successfully test loading daemon, sending a notification, viewing notification, deleting notification.

Based on this test I think it's fit to merge the PR, and I will do the minor work of updating other references to rofication outside of this package( regolith-i3-gaps-config is all I'm aware of but there may be more.)

Impressive effort @farsil , thank you for this awesome contribution!

kgilmer commented:
Have you considered writing a terminal client?  ;)
